token-metadata/program/src/processor/burn/burn.rs

```diff
@@ -1,11 +1,222 @@
+
+    // Assert program ownership.
+    assert_owned_by(ctx.accounts.metadata_info, program_id)?;
+    assert_owned_by(ctx.accounts.mint_info, &spl_token::ID)?;
+    assert_owned_by(ctx.accounts.token_info, &spl_token::ID)?;
+
+    if let Some(edition_info) = ctx.accounts.edition_info {
+        assert_owned_by(edition_info, program_id)?;
+    }
+
+    if let Some(parent_edition) = ctx.accounts.parent_edition_info {
+        assert_owned_by(parent_edition, program_id)?;
+    }
+    if let Some(parent_mint) = ctx.accounts.parent_mint_info {
+        assert_owned_by(parent_mint, &spl_token::ID)?;
+    }
+    if let Some(parent_token) = ctx.accounts.parent_token_info {
+        assert_owned_by(parent_token, &spl_token::ID)?;
+    }
+    if let Some(edition_marker) = ctx.accounts.edition_marker_info {
+        assert_owned_by(edition_marker, program_id)?;
+    }
+
+    // Check program IDs.
+    if ctx.accounts.spl_token_program_info.key != &spl_token::ID {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    if ctx.accounts.system_program_info.key != &system_program::ID {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    if ctx.accounts.sysvar_instructions_info.key != &sysvar::instructions::ID {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    msg!("Deserializing accounts");
+    // Deserialize accounts.
+    let metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
+    let token: TokenAccount = assert_initialized(ctx.accounts.token_info)?;
+
+    msg!("Getting authority type");
+    let authority_response = AuthorityType::get_authority_type(AuthorityRequest {
+        authority: ctx.accounts.authority_info.key,
+        update_authority: &metadata.update_authority,
+        mint: ctx.accounts.mint_info.key,
+        token: Some(ctx.accounts.token_info.key),
+        token_account: Some(&token),
+        token_record_info: ctx.accounts.token_record_info,
+        token_delegate_roles: vec![TokenDelegateRole::Utility],
+        precedence: &[AuthorityType::Holder, AuthorityType::TokenDelegate],
+        ..Default::default()
+    })?;
+
+    match authority_response.authority_type {
+        AuthorityType::Holder => {
+            // Checks:
+            // * Metadata is owned by the token-metadata program
+            // * Mint is owned by the spl-token program
+            // * Token is owned by the spl-token program
+            // * Token account is initialized
+            // * Token account data owner is 'owner'
+            // * Token account belongs to mint
+            // * Token account has 1 or more tokens
+            // * Mint matches metadata.mint
+            assert_currently_holding(
+                &crate::ID,
+                ctx.accounts.authority_info,
+                ctx.accounts.metadata_info,
+                &metadata,
+                ctx.accounts.mint_info,
+                ctx.accounts.token_info,
+            )?;
```

Comment:
Yeah, you're right. I've refactored this to remove duplication and clean this up a bit. See commit [b125b47e9abe197751966729115b292be9af6bdd](https://github.com/metaplex-foundation/metaplex-program-library/pull/1014/commits/b125b47e9abe197751966729115b292be9af6bdd).

Reply:
And commit [849f2e8dde2a617edbf3a4bf72beb4d123ca1548](https://github.com/metaplex-foundation/metaplex-program-library/pull/1014/commits/849f2e8dde2a617edbf3a4bf72beb4d123ca1548) for refactoring the function out of the other legacy burn handlers.

Reply:
Do we still need:
```
if token_account.owner != *owner_info.key {
    return Err(MetadataError::InvalidOwner.into());
}
```
in the case of `AuthorityType::Holder`?

Reply:
The owner check is done in the [`get_authority_type` call](https://github.com/metaplex-foundation/metaplex-program-library/blob/0764b55efca4e54e3ab417f58d466df0e727fcaa/token-metadata/program/src/state/programmable.rs#L383).

> The only benefit of this way is if in some terrible erroneous state token.amount was 2 then it could still burn 1. 

The instruction would still fail later, because it will burn one and then try to close the token account which can't be closed because it still has one token. This is just an invalid state for an NFT and if it ever happened we shouldn't just carry on like it's normal, it should just fail. An NFT with two tokens isn't an NFT and shouldn't work with NFT instructions.